

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Running the Examples &mdash; GEMmaker 1.0 documentation</title>
  

  
  
  
  

  

  
  
    

  

  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="_static/theme_overrides.css" type="text/css" />
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Customizing the Workflow" href="nextflow_config.html" />
    <link rel="prev" title="Software Prerequisites" href="prerequisites.html" /> 

  
  <script src="_static/js/modernizr.min.js"></script>

</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search">
          

          
            <a href="index.html" class="icon icon-home"> GEMmaker
          

          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="prerequisites.html">Software Prerequisites</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Running the Examples</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#analyze-local-data">Analyze Local Data</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#reference-directory">Reference directory</a></li>
<li class="toctree-l3"><a class="reference internal" href="#data-directory">Data directory</a></li>
<li class="toctree-l3"><a class="reference internal" href="#executing-the-example">Executing the example</a></li>
<li class="toctree-l3"><a class="reference internal" href="#results">Results</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#analyze-remote-data">Analyze Remote Data</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id1">Reference directory</a></li>
<li class="toctree-l3"><a class="reference internal" href="#input-file">Input File</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id2">Executing the example</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id3">Results</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#analyzing-both-local-and-remote-together">Analyzing Both Local and Remote Together</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="nextflow_config.html">Customizing the Workflow</a></li>
<li class="toctree-l1"><a class="reference internal" href="preparing_and_running.html">Running your Data</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">GEMmaker</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html">Docs</a> &raquo;</li>
        
      <li>Running the Examples</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="_sources/examples.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="running-the-examples">
<h1>Running the Examples<a class="headerlink" href="#running-the-examples" title="Permalink to this headline">¶</a></h1>
<p>GEMmaker provides two examples for processing RNA-seq samples that are
either stored on a local computer or housed in <a class="reference external" href="https://www.ncbi.nlm.nih.gov/sra">NCBI’s SRA
repository</a>. You can test this
workflow without installing all of the necessary prerequisites by using
the
<a class="reference external" href="https://github.com/SystemsGenetics/GEMmaker-docker">GEMmaker-docker</a>
image.</p>
<div class="section" id="analyze-local-data">
<h2>Analyze Local Data<a class="headerlink" href="#analyze-local-data" title="Permalink to this headline">¶</a></h2>
<p>This example uses the imaginary organism “Cool Organism” (CORG) and a
data set of 3 artificially made RNA-seq runs for CORG. CORG has a very
small “genome” of only 1,953 nucleotides, 2 “chromosomes” and 3 “genes”.
The 3 genes are named “gene_Alpha”, “gene_Beta” and “gene_Zeta”. This
data set is very small so it can be run on a desktop machine in a short
amount of time.</p>
<div class="section" id="reference-directory">
<h3>Reference directory<a class="headerlink" href="#reference-directory" title="Permalink to this headline">¶</a></h3>
<p>The reference directory for the is located at:</p>
<div class="code bash highlight-default"><div class="highlight"><pre><span></span><span class="n">GEMmaker</span><span class="o">/</span><span class="n">examples</span><span class="o">/</span><span class="n">LocalRunExample</span><span class="o">/</span><span class="n">reference</span>
</pre></div>
</div>
<p>This directory contains the</p>
<ul class="simple">
<li>made up reference genome file (<strong>CORG.fna</strong>),</li>
<li><a class="reference external" href="https://uswest.ensembl.org/info/website/upload/gff.html">GTF</a> file (<strong>CORG.gtf</strong>)</li>
<li>hisat index files (<strong>CORG.?/ht2</strong>).</li>
</ul>
</div>
<div class="section" id="data-directory">
<h3>Data directory<a class="headerlink" href="#data-directory" title="Permalink to this headline">¶</a></h3>
<p>The example data directory is located at:</p>
<div class="code bash highlight-default"><div class="highlight"><pre><span></span><span class="o">/</span><span class="n">GEMmaker</span><span class="o">/</span><span class="n">examples</span><span class="o">/</span><span class="n">LocalRunExample</span><span class="o">/</span><span class="n">Data</span><span class="o">/</span>
</pre></div>
</div>
<p>This directory contains 3
<a class="reference external" href="https://en.wikipedia.org/wiki/FASTQ_format">FASTQ</a> files containing
RNA-seq data. These are examples of unpaired data, and are each in a
directory of their own. The file naming format for these reads is
“?_1.fastq” where the “?” is the number of the sample. GEM-maker finds
these files through the glob pattern assigned to the
“local_samples_path” in the <strong>nextflow.config</strong> file.</p>
</div>
<div class="section" id="executing-the-example">
<h3>Executing the example<a class="headerlink" href="#executing-the-example" title="Permalink to this headline">¶</a></h3>
<p>To execute the GEMmaker with the example data you must first rename the
<strong>nextflow.config.example</strong> file as <strong>nextflow.config</strong>:</p>
<div class="code bash highlight-default"><div class="highlight"><pre><span></span><span class="n">mv</span> <span class="n">nextflow</span><span class="o">.</span><span class="n">config</span><span class="o">.</span><span class="n">example</span> <span class="n">nextflow</span><span class="o">.</span><span class="n">config</span>
</pre></div>
</div>
<p>The nextflow.config.example is already configured with the proper
settings to use the local example data just described. To execute the
workflow run the following:</p>
<div class="code bash highlight-default"><div class="highlight"><pre><span></span><span class="n">nextflow</span> <span class="n">run</span> <span class="n">main</span><span class="o">.</span><span class="n">nf</span> <span class="o">-</span><span class="n">profile</span> <span class="n">standard</span>
</pre></div>
</div>
</div>
<div class="section" id="results">
<h3>Results<a class="headerlink" href="#results" title="Permalink to this headline">¶</a></h3>
<p>Once executed, the Local Run Example should output 3 directories. These
will be titled Sample_1, Sample_2, and Sample_3. In a real run,
GEMmaker will automatically combine files that have the same experiment
number( [SED]RX0000000 ) but different run numbers ( [SED]RR0000000 ),
so it is possible that the [SED]RX number contains multiple [SED]RR
runs. However, in the the local example, this is not the case.</p>
<p>In each output directory you will find the following files:</p>
<ul class="simple">
<li><strong>fastq</strong> The fastq reads file for the experiment.</li>
<li><strong>fastqc</strong> 6 or 12 files (depending on paired or unpaired data)
from fastqc. Fastqc is set up tocheck files before and after trimmomatic.</li>
<li><strong>sam</strong> alignment file.</li>
<li><strong>bam</strong> binary alignment file.</li>
<li><strong>ga</strong> expression level transcript abundance.</li>
<li><strong>fpkm</strong> 2 column version of <strong>ga</strong> file with only gene and FPKM value.</li>
<li><strong>tpm</strong> 2 column version of the <strong>ga</strong> file with only gene and TPM values.</li>
</ul>
<p>The output of GEM-maker can be used for several different analysis. The
FPKM files can be combined into an expression matrix and then visualized
using a heatmap. The following heatmap is the Local Example’s fpkm
values divided by 1000 in heatmap form. We can see that gene_Zeta
remained constant across all three samples, gene_Beta decreased, and
gene_Alpha increased.</p>
<div class="figure" id="id4">
<img alt="heatmap" src="_images/heatmap.png" />
<p class="caption"><span class="caption-text">heatmap</span></p>
</div>
</div>
</div>
<div class="section" id="analyze-remote-data">
<h2>Analyze Remote Data<a class="headerlink" href="#analyze-remote-data" title="Permalink to this headline">¶</a></h2>
<p>This example uses data from an unidentified bacteria. The data set is
located on NCBI, and is unusually small, which makes it useful as an
example. This datasset was selected as an example because of it’s small
size which can be run on a desktop computer.</p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">NCBI has all RNA-seq quality score data stored as Phred 33, and by default.
<a class="reference external" href="https://ncbi.github.io/sra-tools/fastq-dump.html">fastq-dump</a> will download
it this way. This is preferable, as Phred33 is the current standard. This
workflow, however, is set up to automatically detect if data is Phred 33 or
64, so it should not matter how quality scores are stored.</p>
</div>
<div class="section" id="id1">
<h3>Reference directory<a class="headerlink" href="#id1" title="Permalink to this headline">¶</a></h3>
<p>The reference directory containing the genome data for the this example
is located at:</p>
<div class="code bash highlight-default"><div class="highlight"><pre><span></span><span class="n">GEMmaker</span><span class="o">/</span><span class="n">examples</span><span class="o">/</span><span class="n">RemoteRunExample</span><span class="o">/</span><span class="n">reference</span>
</pre></div>
</div>
<p>It contains the reference files like the directory for the local
example, but for the unidentified bacteria rather than CORG.</p>
</div>
<div class="section" id="input-file">
<h3>Input File<a class="headerlink" href="#input-file" title="Permalink to this headline">¶</a></h3>
<p>Unlike for local data, analysis of remote data requires a list of <a class="reference external" href="https://www.ncbi.nlm.nih.gov/sra">NCBI
Sequence Read Archive (SRA)</a>. Run
IDs. The file should contain one SRA RUN ID per line, with no blank
lines at the bottom. Here is an example file:</p>
<div class="code bash highlight-default"><div class="highlight"><pre><span></span><span class="n">GEMmaker</span><span class="o">/</span><span class="n">examples</span><span class="o">/</span><span class="n">RemoteRunExample</span><span class="o">/</span><span class="n">SRA_IDs</span><span class="o">.</span><span class="n">txt</span>
</pre></div>
</div>
<p>The file specifies the SRA RUN ID
<a class="reference external" href="https://www.ncbi.nlm.nih.gov/sra/SRR649944/">SRR649944</a></p>
</div>
<div class="section" id="id2">
<h3>Executing the example<a class="headerlink" href="#id2" title="Permalink to this headline">¶</a></h3>
<p>To run the Remote Run Example, you must change some parameters in the
<strong>nextflow.config</strong> file.</p>
<p>First, change the <strong>params.software_params.hisat2.path</strong> so that it
points to the proper reference files directory. For this example, it
should be set to:</p>
<div class="code bash highlight-default"><div class="highlight"><pre><span></span><span class="n">path</span> <span class="o">=</span> <span class="s2">&quot;$</span><span class="si">{PWD}</span><span class="s2">/examples/RemoteRunExample/reference/&quot;</span>
</pre></div>
</div>
<p>The parameter below this is <strong>params.software_params.hisat2.prefix</strong>.
This should be changed to the prefix of the reference files. In this
example, the prefix should be set to:</p>
<div class="code bash highlight-default"><div class="highlight"><pre><span></span><span class="n">prefix</span> <span class="o">=</span> <span class="s2">&quot;GCA_002793175.1_ASM279317v1_genomic&quot;</span>
</pre></div>
</div>
<p>Additionally, <strong>params.input.remote_list_path</strong> should be set to point
to the SRA_IDs.txt file:</p>
<div class="code bash highlight-default"><div class="highlight"><pre><span></span><span class="n">remote_list_path</span> <span class="o">=</span> <span class="s2">&quot;$</span><span class="si">{PWD}</span><span class="s2">/examples/RemoteRunExample/SRA_IDs.txt&quot;</span>
</pre></div>
</div>
<p>The <strong>params.input.local_samples_path</strong> paramter should be set to
indicate that there are no local samples:</p>
<div class="code bash highlight-default"><div class="highlight"><pre><span></span><span class="n">local_samples_path</span> <span class="o">=</span> <span class="s2">&quot;none&quot;</span>
</pre></div>
</div>
</div>
<div class="section" id="id3">
<h3>Results<a class="headerlink" href="#id3" title="Permalink to this headline">¶</a></h3>
<p>Because there is only one smaple, a single directory named “SRR649944”
will be presetn. This is the NCBI experiment ID that the sample belongs
to. It will contain the same output files describe above for the local
run example. It should be noted that this RNA-seq dataset does not
produce any fpkm values.</p>
</div>
</div>
<div class="section" id="analyzing-both-local-and-remote-together">
<h2>Analyzing Both Local and Remote Together<a class="headerlink" href="#analyzing-both-local-and-remote-together" title="Permalink to this headline">¶</a></h2>
<p>GEMmaker is capable of combining both local and remote data in a single
execution if both the <strong>remote_list_path</strong> and
<strong>local_samples_path</strong> are set in the <strong>nextflow.config</strong> file.</p>
</div>
</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="nextflow_config.html" class="btn btn-neutral float-right" title="Customizing the Workflow" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="prerequisites.html" class="btn btn-neutral" title="Software Prerequisites" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  

    
    
      <script type="text/javascript">
          var DOCUMENTATION_OPTIONS = {
              URL_ROOT:'./',
              VERSION:'1.0',
              LANGUAGE:'None',
              COLLAPSE_INDEX:false,
              FILE_SUFFIX:'.html',
              HAS_SOURCE:  true,
              SOURCELINK_SUFFIX: '.txt'
          };
      </script>
        <script type="text/javascript" src="_static/jquery.js"></script>
        <script type="text/javascript" src="_static/underscore.js"></script>
        <script type="text/javascript" src="_static/doctools.js"></script>
    

  

  <script type="text/javascript" src="_static/js/theme.js"></script>

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>