

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Running the Examples &mdash; GEMmaker 1.0 documentation</title>
  

  
  
  
  

  

  
  
    

  

  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="_static/theme_overrides.css" type="text/css" />
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Customizing the Workflow" href="nextflow_config.html" />
    <link rel="prev" title="Software Prerequisites" href="prerequisites.html" /> 

  
  <script src="_static/js/modernizr.min.js"></script>

</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search">
          

          
            <a href="index.html" class="icon icon-home"> GEMmaker
          

          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="prerequisites.html">Software Prerequisites</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Running the Examples</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#executing-the-examples-what-you-came-here-for">Executing the examples (what you came here for)</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#explination-of-the-run">Explination of the run</a></li>
<li class="toctree-l3"><a class="reference internal" href="#running-salmon-or-kallisto-on-the-example-dataset">Running Salmon or Kallisto on the example dataset</a></li>
<li class="toctree-l3"><a class="reference internal" href="#inputs-explained">Inputs explained</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#reference-directory">Reference directory</a></li>
<li class="toctree-l2"><a class="reference internal" href="#data-directory">Data directory</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#outputs-explained">Outputs explained</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#results">Results</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="nextflow_config.html">Customizing the Workflow</a></li>
<li class="toctree-l1"><a class="reference internal" href="preparing_and_running.html">Running your Data</a></li>
<li class="toctree-l1"><a class="reference internal" href="errors_and_example_outputs.html">Error troubleshooting and Example Outputs</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">GEMmaker</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html">Docs</a> &raquo;</li>
        
      <li>Running the Examples</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="_sources/examples.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="running-the-examples">
<span id="id1"></span><h1>Running the Examples<a class="headerlink" href="#running-the-examples" title="Permalink to this headline">¶</a></h1>
<p>The GEMmaker example run will run a small set of local files (downloaded when
you cloned GEMmaker) and a remote file (automatically downloaded from <a class="reference external" href="https://www.ncbi.nlm.nih.gov/sra">NCBI’s
SRA repository</a>.) This demonstrates how
GEMmaker handles local and remote files.</p>
<p>This data set is very small so it can be run on a desktop machine in a short
amount of time. GEMmaker must be cloned to your machine, and it’s
<a class="reference internal" href="prerequisites.html#software-prerequisites"><span class="std std-ref">Software Prerequisites</span></a> must be installed before running these examples.</p>
<div class="section" id="executing-the-examples-what-you-came-here-for">
<h2>Executing the examples (what you came here for)<a class="headerlink" href="#executing-the-examples-what-you-came-here-for" title="Permalink to this headline">¶</a></h2>
<p>Once the <a class="reference internal" href="prerequisites.html#software-prerequisites"><span class="std std-ref">Software Prerequisites</span></a> are installed, you can run
GEMmaker with the example data. If you have not already, you must rename the
<strong>nextflow.config.example</strong> file as <strong>nextflow.config</strong>:</p>
<div class="code bash highlight-default"><div class="highlight"><pre><span></span><span class="n">mv</span> <span class="n">nextflow</span><span class="o">.</span><span class="n">config</span><span class="o">.</span><span class="n">example</span> <span class="n">nextflow</span><span class="o">.</span><span class="n">config</span>
</pre></div>
</div>
<p>The nextflow.config.example is already configured with the proper
settings to run the samples described above. To execute the
workflow run the following:</p>
<div class="code bash highlight-default"><div class="highlight"><pre><span></span><span class="n">nextflow</span> <span class="n">run</span> <span class="n">main</span><span class="o">.</span><span class="n">nf</span> <span class="o">-</span><span class="n">profile</span> <span class="n">standard</span><span class="p">,</span><span class="n">localDocker</span>
</pre></div>
</div>
<p>You should see an output that looks like this: <a class="reference internal" href="errors_and_example_outputs.html#example-output"><span class="std std-ref">Example Output</span></a></p>
<div class="section" id="explination-of-the-run">
<h3>Explination of the run<a class="headerlink" href="#explination-of-the-run" title="Permalink to this headline">¶</a></h3>
<p>This example uses the imaginary organism “Cool Organism” (CORG). For the local
example, we use a set of 3 artificially made RNA-seq runs made for CORG. CORG has a very
small “genome” of only 2,336 nucleotides, 3 “chromosomes” and 6 “genes”.
The 6 genes are named “gene_Alpha”, “gene_Beta”, “gene_Zeta”, “gene_Gamma”,
“gene_Delta”, “gene_Epsilon”.</p>
<p>For the remote example, GEMmaker automatically downloaded a very
small RNA-seq file from NCBI. This dataset is from an uncharacterized bacteria,
but luckly, CORG shares 3 of the genes with this bacteria so we could use CORG’s
reference file (pretend that the remote file is also for CORG, we are just using
it becasue it is an unussually small file, which makes it ideal as an example).</p>
<p>This ran Hisat2. If you would like to run Salmon or Kallisto, please read on.</p>
</div>
<div class="section" id="running-salmon-or-kallisto-on-the-example-dataset">
<h3>Running Salmon or Kallisto on the example dataset<a class="headerlink" href="#running-salmon-or-kallisto-on-the-example-dataset" title="Permalink to this headline">¶</a></h3>
<p>First, you must edit the <code class="docutils literal"><span class="pre">nextflow.config</span></code> file to say that you want to run the
example using Salmon or Kallisto.</p>
<p>In the GEMmaker directiry, edit the <code class="docutils literal"><span class="pre">nextflow.config</span></code> file using your favorite
editor (here we use the <cite>vim &lt;https://www.vim.org/&gt;</cite> editor on the command line,
but you can edit it using a gui based editor if you like. It just needs to be
able to edit a text file)</p>
<p>Then edit the Alignment section of the file. Change to 1 for Kallisto, and 2 for
salmon</p>
<p>Here is what it should look like to run the example with Kallisto</p>
<p>After that, save your file and run the worklow:</p>
<div class="code bash highlight-default"><div class="highlight"><pre><span></span><span class="n">nextflow</span> <span class="n">run</span> <span class="n">main</span><span class="o">.</span><span class="n">nf</span> <span class="o">-</span><span class="n">profile</span> <span class="n">standard</span><span class="p">,</span><span class="n">localDocker</span>
</pre></div>
</div>
</div>
<div class="section" id="inputs-explained">
<h3>Inputs explained<a class="headerlink" href="#inputs-explained" title="Permalink to this headline">¶</a></h3>
<p>The inputs for the example run are in the Reference directory and the 2 data
directories in the examples directory.</p>
</div>
</div>
<div class="section" id="reference-directory">
<h2>Reference directory<a class="headerlink" href="#reference-directory" title="Permalink to this headline">¶</a></h2>
<p>The reference directory for the example is located at:</p>
<div class="code bash highlight-default"><div class="highlight"><pre><span></span><span class="n">GEMmaker</span><span class="o">/</span><span class="n">examples</span><span class="o">/</span><span class="n">reference</span>
</pre></div>
</div>
<p>This directory contains the</p>
<ul class="simple">
<li>made up reference genome file (<strong>CORG.fna</strong>),</li>
<li><a class="reference external" href="https://uswest.ensembl.org/info/website/upload/gff.html">GTF</a> file (<strong>CORG.gtf</strong>)</li>
<li>hisat index files (<strong>CORG.?/ht2</strong>).</li>
<li>kallisto index file (<strong>CORG.transcripts.Kallisto.indexed</strong>)</li>
<li>salmon index directory (<strong>CORG.transcripts.Salmon.indexed/</strong>)</li>
<li><strong>COMMANDS.sh</strong> explaining how each of these were generated</li>
</ul>
<p>These are the files needed to run hisat2, kallisto, and salmon on the CORG data.</p>
</div>
<div class="section" id="data-directory">
<h2>Data directory<a class="headerlink" href="#data-directory" title="Permalink to this headline">¶</a></h2>
<p>There are 2 sample data directories:</p>
<p>For local runs:</p>
<div class="code bash highlight-default"><div class="highlight"><pre><span></span><span class="o">/</span><span class="n">GEMmaker</span><span class="o">/</span><span class="n">examples</span><span class="o">/</span><span class="n">LocalRunExample</span><span class="o">/</span><span class="n">Data</span><span class="o">/</span>
</pre></div>
</div>
<p>For remote runs:</p>
<div class="code bash highlight-default"><div class="highlight"><pre><span></span><span class="o">/</span><span class="n">GEMmaker</span><span class="o">/</span><span class="n">examples</span><span class="o">/</span><span class="n">RemoteRunExample</span><span class="o">/</span>
</pre></div>
</div>
<p>The Local runs directory contains 3 <a class="reference external" href="https://en.wikipedia.org/wiki/FASTQ_format">FASTQ</a> files for CORG containing
RNA-seq data. These are examples of local unpaired data, and are each in a directory
of their own. The file naming format for these reads is “?_1.fastq” where the “?”
is the number of the sample. GEM-maker finds these files through the glob
pattern assigned to the “local_samples_path” in the <strong>nextflow.config</strong> file.</p>
<p>The Remote runs directory contains the file <code class="docutils literal"><span class="pre">SRA_IDs.txt</span></code> which contains the
name of the remote file to be downloaded by GEMmaker from <a class="reference external" href="https://www.ncbi.nlm.nih.gov/sra">NCBI’s SRA repository</a>. In a real run, this would probably contain
more than just one sra name, with each run being on a new line.</p>
<div class="section" id="outputs-explained">
<h3>Outputs explained<a class="headerlink" href="#outputs-explained" title="Permalink to this headline">¶</a></h3>
</div>
</div>
<div class="section" id="results">
<h2>Results<a class="headerlink" href="#results" title="Permalink to this headline">¶</a></h2>
<p>Once executed, the example should output a directory title <code class="docutils literal"><span class="pre">output</span></code> with 6 directories in it.
Four of the directories correspond to each the samples run (3 local, 1 remote). Each
of these contains the files generated for each of these samples.</p>
<p>In each output directory you will find the following files:</p>
<ul class="simple">
<li><strong>fastq</strong> The fastq reads file for the experiment.</li>
<li><strong>fastqc</strong> 6 or 12 files (depending on paired or unpaired data)
from fastqc. Fastqc is set up tocheck files before and after trimmomatic.</li>
<li><strong>sam</strong> alignment file.</li>
<li><strong>bam</strong> binary alignment file.</li>
<li><strong>ga</strong> expression level transcript abundance.</li>
<li><strong>fpkm</strong> 2 column version of <strong>ga</strong> file with only gene and FPKM value.</li>
<li><strong>tpm</strong> 2 column version of the <strong>ga</strong> file with only gene and TPM values.</li>
</ul>
<p>The remaining 2 directories are the <code class="docutils literal"><span class="pre">reports</span></code> directory and the <code class="docutils literal"><span class="pre">GEM</span></code>
directory. The <code class="docutils literal"><span class="pre">reports</span></code> directory will contain a <code class="docutils literal"><span class="pre">multiqc_report.html</span></code> file
that reports on different statistics about the run.</p>
<div class="figure">
<img alt="MultiQC_Report" src="_images/MultiQC_Report.png" />
</div>
<p>Figure 1: Image of the start of the report for the example run when run with Hisat2.</p>
<p>The <code class="docutils literal"><span class="pre">GEM</span></code> directory contains the final GEM matrix, in raw, TPM and FPKM form.
These can be used for further analysis.</p>
<p>The output of GEM-maker can be used for several different analysis. The
FPKM files can be combined into an expression matrix and then visualized
using a heatmap. The following heatmap is the Local Example’s fpkm
values divided by 1000 in heatmap form. We can see that gene_Zeta
remained constant across all three samples, gene_Beta decreased, and
gene_Alpha increased.</p>
<div class="figure" id="id2">
<img alt="heatmap" src="_images/heatmap.png" />
<p class="caption"><span class="caption-text">heatmap</span></p>
</div>
</div>
</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="nextflow_config.html" class="btn btn-neutral float-right" title="Customizing the Workflow" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="prerequisites.html" class="btn btn-neutral" title="Software Prerequisites" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  

    
    
      <script type="text/javascript">
          var DOCUMENTATION_OPTIONS = {
              URL_ROOT:'./',
              VERSION:'1.0',
              LANGUAGE:'None',
              COLLAPSE_INDEX:false,
              FILE_SUFFIX:'.html',
              HAS_SOURCE:  true,
              SOURCELINK_SUFFIX: '.txt'
          };
      </script>
        <script type="text/javascript" src="_static/jquery.js"></script>
        <script type="text/javascript" src="_static/underscore.js"></script>
        <script type="text/javascript" src="_static/doctools.js"></script>
    

  

  <script type="text/javascript" src="_static/js/theme.js"></script>

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>