

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Customizing the Workflow &mdash; GEMmaker 1.0 documentation</title>
  

  
  
  
  

  

  
  
    

  

  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="_static/theme_overrides.css" type="text/css" />
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Running your Data" href="preparing_and_running.html" />
    <link rel="prev" title="Running the Examples" href="examples.html" /> 

  
  <script src="_static/js/modernizr.min.js"></script>

</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search">
          

          
            <a href="index.html" class="icon icon-home"> GEMmaker
          

          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="prerequisites.html">Software Prerequisites</a></li>
<li class="toctree-l1"><a class="reference internal" href="examples.html">Running the Examples</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Customizing the Workflow</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#input">Input</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#remote-list-path">remote_list_path</a></li>
<li class="toctree-l3"><a class="reference internal" href="#local-samples-path">local_samples_path</a></li>
<li class="toctree-l3"><a class="reference internal" href="#reference-path">reference_path</a></li>
<li class="toctree-l3"><a class="reference internal" href="#reference-prefix">reference_prefix</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#output">Output</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#dir">dir</a></li>
<li class="toctree-l3"><a class="reference internal" href="#sample-dir">sample_dir</a></li>
<li class="toctree-l3"><a class="reference internal" href="#publish-mode">publish_mode</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#execution">Execution</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#queue-size">queue_size</a></li>
<li class="toctree-l3"><a class="reference internal" href="#threads">threads</a></li>
<li class="toctree-l3"><a class="reference internal" href="#max-retries">max_retries</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="preparing_and_running.html">Running your Data</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">GEMmaker</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html">Docs</a> &raquo;</li>
        
      <li>Customizing the Workflow</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="_sources/nextflow_config.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="customizing-the-workflow">
<h1>Customizing the Workflow<a class="headerlink" href="#customizing-the-workflow" title="Permalink to this headline">¶</a></h1>
<p>All nextflow workflows have a <code class="docutils literal"><span class="pre">nextflow.config</span></code> file. It is where parameters for
the workflow can be customized.</p>
<p>For GEMmaker, the <code class="docutils literal"><span class="pre">nextflow.config</span></code> file is in 3 main parts:</p>
<ul class="simple">
<li>params - the largest section, contains information on the input files, output
files and software</li>
<li>performance - contains sections on the performance of the workflow</li>
<li>profiles - nextflow can run on multiple different HPC systems, this is
where you modify this parameter. Refer to the <a class="reference external" href="https://www.nextflow.io/docs/latest/config.html#config-profiles">Nextflow
documentation</a>
to modify the profile for your system</li>
</ul>
<p>The following gives detailed information on each parameter in the
<code class="docutils literal"><span class="pre">nextflow.config</span></code> file.</p>
<div class="section" id="input">
<h2>Input<a class="headerlink" href="#input" title="Permalink to this headline">¶</a></h2>
<div class="section" id="remote-list-path">
<h3>remote_list_path<a class="headerlink" href="#remote-list-path" title="Permalink to this headline">¶</a></h3>
<p>The path (full or relative) to the list of fastq_ids to be downloaded
from NCBI. This must be a text file with one SRR/DRR/ERR number per line.
No blank lines are allowed
If no remote files are to be downloaded, set parameter as “none”</p>
<p>Default:</p>
<div class="code bash highlight-default"><div class="highlight"><pre><span></span><span class="n">remote_list_path</span> <span class="o">=</span> <span class="s2">&quot;none&quot;</span>
</pre></div>
</div>
</div>
<div class="section" id="local-samples-path">
<h3>local_samples_path<a class="headerlink" href="#local-samples-path" title="Permalink to this headline">¶</a></h3>
<p>The glob that retrieves locally stored “fastq” files
An example of a proper glob to retreive files can be seen below as the
default. This glob pattern will find all files that have an ending of
“_1.fastq” or “_2.fastq” in the subdirectories of the folder “Sample”.
If no local files are to be used, set parameter as “none”</p>
<p>Default:</p>
<div class="code bash highlight-default"><div class="highlight"><pre><span></span><span class="n">local_samples_path</span> <span class="o">=</span> <span class="s2">&quot;$</span><span class="si">{PWD}</span><span class="s2">/examples/LocalRunExample/Data/Sample*/*_{1,2}.fastq&quot;</span>
</pre></div>
</div>
</div>
<div class="section" id="reference-path">
<h3>reference_path<a class="headerlink" href="#reference-path" title="Permalink to this headline">¶</a></h3>
<p>The full file system path of the directory containing the genome reference
files. The reference genome is provided to this workflow via a set of files
in a single directory. The list of files includes:</p>
<ol class="arabic simple">
<li>A FASTA file containing the genomic sequence</li>
<li>Hisat2 index files for the reference, which must be created with bisat2-bulid</li>
<li>A GTF file containing the genes annotated within hte genome</li>
</ol>
<p>All files for the reference genome must begin with the same file prefix. For
example, if the prefix is TAIR10_Araport11 then the following files should be
present all with “TAIR10_Araport11” as the file prefix:
TAIR10_Araport11.fna, TAIR10_Araport11.1.ht2, TAIR10_Araport11.2.ht2, (with
potentially more hisat2 index files), and TAIR10_Araport11.gtf.</p>
<p>Default:</p>
<div class="code bash highlight-default"><div class="highlight"><pre><span></span><span class="n">reference_path</span> <span class="o">=</span> <span class="s2">&quot;$</span><span class="si">{PWD}</span><span class="s2">/examples/LocalRunExample/reference/&quot;</span>
</pre></div>
</div>
</div>
<div class="section" id="reference-prefix">
<h3>reference_prefix<a class="headerlink" href="#reference-prefix" title="Permalink to this headline">¶</a></h3>
<p>The prefix (used by hisat2-build) for the genome reference files. Note:
all files in the reference directory must have this prefix as well.</p>
<p>Default:</p>
<div class="code bash highlight-default"><div class="highlight"><pre><span></span><span class="nb">dir</span> <span class="o">=</span> <span class="s2">&quot;$</span><span class="si">{PWD}</span><span class="s2">/output&quot;</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="output">
<h2>Output<a class="headerlink" href="#output" title="Permalink to this headline">¶</a></h2>
<div class="section" id="dir">
<h3>dir<a class="headerlink" href="#dir" title="Permalink to this headline">¶</a></h3>
<p>All results and reports generated by nextflow are stored in a single
output directory to make it easier to transfer results to long-term
storage such as iRODS.</p>
<p>Default:</p>
<div class="code bash highlight-default"><div class="highlight"><pre><span></span><span class="nb">dir</span> <span class="o">=</span> <span class="s2">&quot;$</span><span class="si">{PWD}</span><span class="s2">/output&quot;</span>
</pre></div>
</div>
</div>
<div class="section" id="sample-dir">
<h3>sample_dir<a class="headerlink" href="#sample-dir" title="Permalink to this headline">¶</a></h3>
<p>Results generated by this workflow are stored in directories that use
“sample_id”. as directory name. If the “fastq_run_id” is not associated
with a “sample_id” (for example, with local files), then a “sample_id”
will be automatically assigned by adding “<a href="#id1"><span class="problematic" id="id2">Sample_</span></a>” to the begining of the
“fastq_run_id” (for example, “123_file1_1.fastq” would be assigned the
sample_id “Sample_123_file1_1”). The default storage pattern is to make
one directory for each “sample_id”, with the parameter set as:</p>
<blockquote>
<div><div class="code bash highlight-default"><div class="highlight"><pre><span></span><span class="n">sample_dir</span> <span class="o">=</span> <span class="p">{</span> <span class="s2">&quot;$</span><span class="si">{params.output.dir}</span><span class="s2">/$</span><span class="si">{sample_id}</span><span class="s2">&quot;</span> <span class="p">}</span>
</pre></div>
</div>
</div></blockquote>
<p>However, if you have a large amount of samples (typically 1000 +), it may
be problematic to have hundreds or thousands of sample directories in
one place. To fix this you can assign a glob pattern to organize the
results into a cascading file system. For example, the following:</p>
<blockquote>
<div><div class="code bash highlight-default"><div class="highlight"><pre><span></span><span class="n">sample_dir</span> <span class="o">=</span> <span class="p">{</span> <span class="s2">&quot;$</span><span class="si">{params.output.dir}</span><span class="s2">/$</span><span class="si">{sample_id[0..2]}</span><span class="s2">/$</span><span class="si">{sample_id[3..4]}</span><span class="s2">/${sample_id.drop(5)}/$</span><span class="si">{sample_id}</span><span class="s2">&quot;</span> <span class="p">}</span>
</pre></div>
</div>
</div></blockquote>
<p>Will organize files downloaded from NCBI in a nesting fashion. The
output of the sample_id “SRX0123456” would be put in the directory
“/SRX/12/34/56/SRX123456/”. You can modify the above glob patterns for
your needs.</p>
<p>Default:</p>
<div class="code bash highlight-default"><div class="highlight"><pre><span></span><span class="n">sample_dir</span> <span class="o">=</span> <span class="p">{</span> <span class="s2">&quot;$</span><span class="si">{params.output.dir}</span><span class="s2">/$</span><span class="si">{sample_id}</span><span class="s2">&quot;</span> <span class="p">}</span>
</pre></div>
</div>
</div>
<div class="section" id="publish-mode">
<h3>publish_mode<a class="headerlink" href="#publish-mode" title="Permalink to this headline">¶</a></h3>
<p>publish mode for publishDir
Options are the standard nextflow stage options. These are:</p>
<ul class="simple">
<li><code class="docutils literal"><span class="pre">&quot;link&quot;</span></code>     Reccomended, this creates a hardlink of all files that should be published</li>
<li><code class="docutils literal"><span class="pre">&quot;symlink&quot;</span></code>  Use when hardlink is not possible.</li>
<li><code class="docutils literal"><span class="pre">&quot;copy&quot;</span></code>     Not recommended. This copies all files to desired publshdir
after they are created in the pipeline. Takes alot of memory and
will slow down pipeline drastically</li>
</ul>
<p>Default:</p>
<div class="code bash highlight-default"><div class="highlight"><pre><span></span><span class="n">publish_mode</span> <span class="o">=</span> <span class="s2">&quot;link&quot;</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="execution">
<h2>Execution<a class="headerlink" href="#execution" title="Permalink to this headline">¶</a></h2>
<div class="section" id="queue-size">
<h3>queue_size<a class="headerlink" href="#queue-size" title="Permalink to this headline">¶</a></h3>
<p>Maximum number of processes to execute at once</p>
<p>Default:</p>
<div class="code bash highlight-default"><div class="highlight"><pre><span></span><span class="n">queue_size</span> <span class="o">=</span> <span class="mi">100</span>
</pre></div>
</div>
</div>
<div class="section" id="threads">
<h3>threads<a class="headerlink" href="#threads" title="Permalink to this headline">¶</a></h3>
<p>Number of threads for multi-threaded processes</p>
<p>Default:</p>
<div class="code bash highlight-default"><div class="highlight"><pre><span></span><span class="n">threads</span> <span class="o">=</span> <span class="mi">1</span>
</pre></div>
</div>
</div>
<div class="section" id="max-retries">
<h3>max_retries<a class="headerlink" href="#max-retries" title="Permalink to this headline">¶</a></h3>
<p>Error strategy for when a process fails <cite>max_retries</cite> times. Can
be <code class="docutils literal"><span class="pre">&quot;terminate&quot;</span></code>, <code class="docutils literal"><span class="pre">&quot;finish&quot;</span></code>, or <code class="docutils literal"><span class="pre">&quot;ignore&quot;</span></code>.</p>
<p>Default:</p>
<div class="code bash highlight-default"><div class="highlight"><pre><span></span><span class="n">max_retries</span> <span class="o">=</span> <span class="s2">&quot;ignore&quot;</span>
</pre></div>
</div>
</div>
</div>
</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="preparing_and_running.html" class="btn btn-neutral float-right" title="Running your Data" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="examples.html" class="btn btn-neutral" title="Running the Examples" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  

    
    
      <script type="text/javascript">
          var DOCUMENTATION_OPTIONS = {
              URL_ROOT:'./',
              VERSION:'1.0',
              LANGUAGE:'None',
              COLLAPSE_INDEX:false,
              FILE_SUFFIX:'.html',
              HAS_SOURCE:  true,
              SOURCELINK_SUFFIX: '.txt'
          };
      </script>
        <script type="text/javascript" src="_static/jquery.js"></script>
        <script type="text/javascript" src="_static/underscore.js"></script>
        <script type="text/javascript" src="_static/doctools.js"></script>
    

  

  <script type="text/javascript" src="_static/js/theme.js"></script>

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>