

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Running your Data &mdash; GEMmaker 1.0 documentation</title>
  

  
  
  
  

  

  
  
    

  

  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="_static/theme_overrides.css" type="text/css" />
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="prev" title="Customizing the Workflow" href="nextflow_config.html" /> 

  
  <script src="_static/js/modernizr.min.js"></script>

</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search">
          

          
            <a href="index.html" class="icon icon-home"> GEMmaker
          

          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="prerequisites.html">Software Prerequisites</a></li>
<li class="toctree-l1"><a class="reference internal" href="examples.html">Running the Examples</a></li>
<li class="toctree-l1"><a class="reference internal" href="nextflow_config.html">Customizing the Workflow</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Running your Data</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#step-1-rna-seq-sample-location">Step 1) RNA-seq sample location</a></li>
<li class="toctree-l2"><a class="reference internal" href="#step-2-aquire-reference-genome-files">Step 2) Aquire Reference Genome Files</a></li>
<li class="toctree-l2"><a class="reference internal" href="#step-3-executing-the-workflow">Step 3) Executing the Workflow</a></li>
<li class="toctree-l2"><a class="reference internal" href="#performance-considerations">Performance Considerations</a></li>
<li class="toctree-l2"><a class="reference internal" href="#generating-a-summary-report">Generating a Summary Report</a></li>
<li class="toctree-l2"><a class="reference internal" href="#generating-the-gene-expression-matrix-gem">Generating the Gene Expression Matrix (GEM)</a></li>
<li class="toctree-l2"><a class="reference internal" href="#using-the-gem-in-other-analysis">Using the GEM in other Analysis</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#dge-analysis">DGE Analysis</a></li>
<li class="toctree-l3"><a class="reference internal" href="#network-analysis">Network Analysis</a></li>
</ul>
</li>
</ul>
</li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">GEMmaker</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html">Docs</a> &raquo;</li>
        
      <li>Running your Data</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="_sources/preparing_and_running.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="running-your-data">
<h1>Running your Data<a class="headerlink" href="#running-your-data" title="Permalink to this headline">¶</a></h1>
<div class="section" id="step-1-rna-seq-sample-location">
<h2>Step 1) RNA-seq sample location<a class="headerlink" href="#step-1-rna-seq-sample-location" title="Permalink to this headline">¶</a></h2>
<p>GEMmaker supports processing of sample files that are already present on
your local computer or samples that are stored in the <a class="reference external" href="https://www.ncbi.nlm.nih.gov/sra">NCBI SRA
repository</a>.</p>
<ul class="simple">
<li>For local sample files, identify a <a class="reference external" href="https://en.wikipedia.org/wiki/Glob_(programming)">glob
pattern</a> that
finds these files.</li>
<li>For samples on NCBI, identify the NCBI SRA run IDs of the samples you
want to analyze. The Run IDs typically start with an SRR, ERR or DRR
prefix. hese sample run IDs must be placed, one per line, in a file
and the filename should be set in the <code class="docutils literal"><span class="pre">remote_list_path</span></code> of
<code class="docutils literal"><span class="pre">nextflow.config</span></code>.</li>
</ul>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">The SRA Toolkit caches SRA files in your home directory by
default. For large experiments this cache can become quite large, which
may become an issue on some HPC clusters where each user is given a disk
quota for their home directory. You can change the location of the cache
by running <code class="docutils literal"><span class="pre">vdb-config</span> <span class="pre">-i</span></code> (see <a class="reference external" href="https://github.com/ncbi/sra-tools/wiki/Toolkit-Configuration">SRA Toolkit
Configuration</a>).</p>
</div>
</div>
<div class="section" id="step-2-aquire-reference-genome-files">
<h2>Step 2) Aquire Reference Genome Files<a class="headerlink" href="#step-2-aquire-reference-genome-files" title="Permalink to this headline">¶</a></h2>
<p>Download the genome annotation/reference files. These will differ depending on
your organism that you are performing the analysis on. You must have the
following:</p>
<ul>
<li><p class="first">A FASTA file containing the full genomic sequence (either
pseudomolecules or scaffolds). Note, if your genome file is extremely
large with hundreds of thousands of contigs/scaffolds, you may want
to reduce the size of the FASTA file to contain only those
contigs/scaffolds with predicted annotated genes.</p>
</li>
<li><p class="first">A <a class="reference external" href="https://uswest.ensembl.org/info/website/upload/gff.html">GTF</a>
file containing the gene models.</p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p>Sometimes a genome assembly does not
provide a GTF file, but rather provides a
<a class="reference external" href="https://uswest.ensembl.org/info/website/upload/gff.html">GFF3</a>
file. You can convert the GFF file to a GTF file using the
<code class="docutils literal"><span class="pre">gffread</span></code> program of
<a class="reference external" href="http://cole-trapnell-lab.github.io/cufflinks/file_formats/">cufflinks</a>,
which you may have to download separately. An example command-line to
convert a GFF3 to GTF is</p>
<div class="code bash highlight-default"><div class="highlight"><pre><span></span><span class="n">gffread</span> <span class="p">[</span><span class="n">gff_file</span><span class="p">]</span> <span class="o">-</span><span class="n">T</span> <span class="o">-</span><span class="n">o</span> <span class="p">[</span><span class="n">gtf_file</span><span class="p">]</span>
</pre></div>
</div>
<p class="last">where <code class="docutils literal"><span class="pre">[gff_file]</span></code> and <code class="docutils literal"><span class="pre">[gtf_file]</span></code> should be substituted for the
names of your GFF3 and desired GTF file respectively.</p>
</div>
</li>
</ul>
<p>Additional considerations:</p>
<ul>
<li><p class="first">You must have hisat2 index files of your genome sequence. These are
constructed by using the <code class="docutils literal"><span class="pre">hisat2-build</span></code> command.</p>
<p>hisat2-build -f YourFastaFile.fna YourPrefix | tee &gt; hisat2-build.log</p>
</li>
<li><p class="first">The GTF file and the hisat2 index files must have the same prefix and
this prefix must be identified in <code class="docutils literal"><span class="pre">nextflow.config</span></code> using the
<code class="docutils literal"><span class="pre">prefix</span></code> parameter for <code class="docutils literal"><span class="pre">hisat2-build</span></code>.</p>
<p>CORG.1.ht2</p>
</li>
</ul>
<p>CORG.2.ht2
CORG.3.ht2
CORG.4.ht2
CORG.5.ht2
CORG.6.ht2
CORG.7.ht2
CORG.8.ht2
CORG.fna
CORG.gtf</p>
<ul class="simple">
<li>All of the genome annotation files must be in a directory and this
directory must be identified in <code class="docutils literal"><span class="pre">nextflow.config</span></code> using the
<code class="docutils literal"><span class="pre">ref</span> <span class="pre">&gt;</span> <span class="pre">path</span></code> paramter.</li>
</ul>
<p>The GEMmaker repo contains an <code class="docutils literal"><span class="pre">examples</span></code> directory which contains
several small example setups. The <code class="docutils literal"><span class="pre">LocalRunExample</span></code> contains a
<code class="docutils literal"><span class="pre">reference</span></code> directory whose files have the same prefix of <code class="docutils literal"><span class="pre">CORG</span></code>.
This prefix is set for the <code class="docutils literal"><span class="pre">prefix</span></code> parameter in
<code class="docutils literal"><span class="pre">nextflow.config.example</span></code>. The <code class="docutils literal"><span class="pre">RemoteRunExample</span></code> also contains an
<code class="docutils literal"><span class="pre">SRA_IDS.txt</span></code> file which contains a list of SRA fastq_run_IDs to
download from NCBI.</p>
</div>
<div class="section" id="step-3-executing-the-workflow">
<h2>Step 3) Executing the Workflow<a class="headerlink" href="#step-3-executing-the-workflow" title="Permalink to this headline">¶</a></h2>
<p>To execute the workflow on a local machine:</p>
<div class="code bash highlight-default"><div class="highlight"><pre><span></span><span class="n">nextflow</span> <span class="n">run</span> <span class="n">main</span><span class="o">.</span><span class="n">nf</span> <span class="o">-</span><span class="n">profile</span> <span class="n">standard</span>
</pre></div>
</div>
<p>To resume a workflow in the event of a failure:</p>
<div class="code bash highlight-default"><div class="highlight"><pre><span></span><span class="n">nextflow</span> <span class="n">run</span> <span class="n">main</span><span class="o">.</span><span class="n">nf</span> <span class="o">-</span><span class="n">profile</span> <span class="n">standard</span> <span class="o">-</span><span class="n">resume</span>
</pre></div>
</div>
<p>To execute the workflow and generate trace, timeline and execution
reports:</p>
<div class="code bash highlight-default"><div class="highlight"><pre><span></span><span class="n">nextflow</span> <span class="n">run</span> <span class="n">main</span><span class="o">.</span><span class="n">nf</span> <span class="o">-</span><span class="n">profile</span> <span class="n">standard</span> <span class="o">-</span><span class="k">with</span><span class="o">-</span><span class="n">report</span> <span class="o">-</span><span class="k">with</span><span class="o">-</span><span class="n">timeline</span> <span class="o">-</span><span class="k">with</span><span class="o">-</span><span class="n">trace</span>
</pre></div>
</div>
<p>To execute the workflow on an HPC system you must edit
<code class="docutils literal"><span class="pre">nextflow.config</span></code> and add an appropriate profile for your system.
Refer to the <a class="reference external" href="https://www.nextflow.io/docs/latest/config.html#config-profiles">Nextflow
documentation</a>.
You can then use any of the above commands by changing the <code class="docutils literal"><span class="pre">-profile</span></code>
argument to use your profile.</p>
</div>
<div class="section" id="performance-considerations">
<h2>Performance Considerations<a class="headerlink" href="#performance-considerations" title="Permalink to this headline">¶</a></h2>
<p>For large experiments on an HPC system, it is important to make sure
that you are effectively utilizing the resources of the system. There
are a number of parameters in <code class="docutils literal"><span class="pre">nextflow.config</span></code> which can be used to
increase performance based on the capabilities of your system: -
<code class="docutils literal"><span class="pre">params.execution.threads</span></code>: All processes which support multithreading
(such as trimmomatic) will use this number of threads. This setting
should be determined by the number of cores per node on your system; for
example, if your system has nodes with 16 cores per node then you could
set the number of threads to 16 to make full use of those nodes. -
<code class="docutils literal"><span class="pre">params.execution.queue_size</span></code>: Nextflow will only run 100 processes at
a time by default, but you may be able to increase this value based on
the queue limits of your system. - <code class="docutils literal"><span class="pre">scratch</span></code>: If this directive is
enabled in your profile, each process will attempt to use its own local
disk space instead of the main working directory. If you are running the
workflow from NFS storage then you may benefit greatly from this option,
if the nodes on your system have sufficient local storage. This option
can significantly reduce the amount of disk I/O on your NFS storage,
since each process will only interact with the NFS storage at the
beginning and end to transfer input and output files.</p>
</div>
<div class="section" id="generating-a-summary-report">
<h2>Generating a Summary Report<a class="headerlink" href="#generating-a-summary-report" title="Permalink to this headline">¶</a></h2>
<p>The <a class="reference external" href="http://multiqc.info">MultiQC</a> tool can be used with GEMmaker to
generate a summary report of results from Trimmomatic, Hisat2 and
samtools. This report allows you to explore the quality of the data,
trimming and alignments. To generate the report you must have <a class="reference external" href="http://multiqc.info/docs/#installing-multiqc">MultiQC
installed</a>. Once
installed, you can generate the report with the following command inside
of the GEMmaker directory where your workflow was executed:</p>
<div class="code bash highlight-default"><div class="highlight"><pre><span></span><span class="n">multiqc</span> <span class="o">.</span>
</pre></div>
</div>
</div>
<div class="section" id="generating-the-gene-expression-matrix-gem">
<h2>Generating the Gene Expression Matrix (GEM)<a class="headerlink" href="#generating-the-gene-expression-matrix-gem" title="Permalink to this headline">¶</a></h2>
<p>After GEMmaker completes, the results for each sample are stored in a
directory specific to that sample. The final output for each sample is a
Gene Expression Vector (GEV) in the form of an FPKM or TPM file. To
compile all GEVs into a Gene Expression Matrix (GEM) you can use the
<code class="docutils literal"><span class="pre">create_GEM.py</span></code> script in the <code class="docutils literal"><span class="pre">scripts</span></code> directory.</p>
<p>To see help documentation for this script:</p>
<div class="code bash highlight-default"><div class="highlight"><pre><span></span><span class="n">python</span> <span class="o">./</span><span class="n">scripts</span><span class="o">/</span><span class="n">create_GEM</span><span class="o">.</span><span class="n">py</span> <span class="o">-</span><span class="n">h</span>
</pre></div>
</div>
<p>To create a GEM file from the TPM files produced by GEMmaker:</p>
<div class="code bash highlight-default"><div class="highlight"><pre><span></span><span class="n">python</span> <span class="o">./</span><span class="n">scripts</span><span class="o">/</span><span class="n">create_GEM</span><span class="o">.</span><span class="n">py</span> <span class="o">--</span><span class="n">source</span> <span class="o">./</span> <span class="o">--</span><span class="nb">type</span> <span class="n">TPM</span> <span class="o">--</span><span class="n">prefix</span> <span class="n">my_project</span>
</pre></div>
</div>
<p>The script will produce a GEM file called <code class="docutils literal"><span class="pre">my_project.GEM.TPM.txt</span></code>. Be
sure to change <code class="docutils literal"><span class="pre">my_project</span></code> to a meaningful prefix for your project.</p>
<p>You can combine the results of multiple GEMmaker runs into a single GEM
by providing a list of directories to the <code class="docutils literal"><span class="pre">--source</span></code> argument. This
feature may be useful if you split a set of input files into several
GEMmaker runs and now you need to combine then. The script will produce
a file named <code class="docutils literal"><span class="pre">GEM.txt</span></code> in the working directory.</p>
</div>
<div class="section" id="using-the-gem-in-other-analysis">
<h2>Using the GEM in other Analysis<a class="headerlink" href="#using-the-gem-in-other-analysis" title="Permalink to this headline">¶</a></h2>
<div class="section" id="dge-analysis">
<h3>DGE Analysis<a class="headerlink" href="#dge-analysis" title="Permalink to this headline">¶</a></h3>
<p>Need to do research on:</p>
</div>
<div class="section" id="network-analysis">
<h3>Network Analysis<a class="headerlink" href="#network-analysis" title="Permalink to this headline">¶</a></h3>
<p>After construction of the GEM, network analysis can be performed.
<a class="reference external" href="https://github.com/SystemsGenetics/KINC">KINC</a> (Knowledge Independent
Network Construction) is a high performance gene co-expression  that can perform
Pearson’s or Spearman’s correlation with K-means or Gaussian mixture models. KINC
is a Qt/<a class="reference external" href="https://github.com/SystemsGenetics/ACE">ACE</a> application that is
capable of running on GPU’s, making it fast and efficient.</p>
</div>
</div>
</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
      
        <a href="nextflow_config.html" class="btn btn-neutral" title="Customizing the Workflow" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  

    
    
      <script type="text/javascript">
          var DOCUMENTATION_OPTIONS = {
              URL_ROOT:'./',
              VERSION:'1.0',
              LANGUAGE:'None',
              COLLAPSE_INDEX:false,
              FILE_SUFFIX:'.html',
              HAS_SOURCE:  true,
              SOURCELINK_SUFFIX: '.txt'
          };
      </script>
        <script type="text/javascript" src="_static/jquery.js"></script>
        <script type="text/javascript" src="_static/underscore.js"></script>
        <script type="text/javascript" src="_static/doctools.js"></script>
    

  

  <script type="text/javascript" src="_static/js/theme.js"></script>

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>